File: android-native/feature/auth/src/main/kotlin/today/mindlog/todo/feature/auth/LoginViewModel.kt
package today.mindlog.todo.feature.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import today.mindlog.todo.core.data.AuthProviders
import today.mindlog.todo.core.data.AuthRepository
import today.mindlog.todo.core.data.OAuthCallback
import today.mindlog.todo.core.data.OAuthCallbackBus
import javax.inject.Inject

data class LoginUiState(
    val email: String = "",
    val password: String = "",
    val busy: Boolean = false,
    val error: String? = null,
    val providers: AuthProviders = AuthProviders(),
    /** Non-null once mindlog id returned an account without an email. */
    val pendingToken: String? = null,
    val pendingEmail: String = "",
) {
    val canSubmit: Boolean get() = !busy && email.isNotBlank() && password.isNotBlank()
}

@HiltViewModel
class LoginViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val callbackBus: OAuthCallbackBus,
) : ViewModel() {

    private val _state = MutableStateFlow(LoginUiState())
    val state: StateFlow<LoginUiState> = _state.asStateFlow()

    init {
        // Ask the server which sign-in paths it has configured before drawing
        // any of their buttons.
        viewModelScope.launch {
            _state.update { it.copy(providers = authRepository.authProviders()) }
        }
        viewModelScope.launch {
            callbackBus.callbacks.collect(::onOAuthCallback)
        }
    }

    fun onEmailChange(value: String) = _state.update { it.copy(email = value, error = null) }
    fun onPasswordChange(value: String) = _state.update { it.copy(password = value, error = null) }
    fun onPendingEmailChange(value: String) = _state.update { it.copy(pendingEmail = value) }

    fun signIn() {
        val current = _state.value
        if (!current.canSubmit) return
        _state.update { it.copy(busy = true, error = null) }
        viewModelScope.launch {
            authRepository.login(current.email.trim(), current.password)
                .onFailure { e -> _state.update { it.copy(busy = false, error = e.readable()) } }
            // On success SessionStore flips `signedIn` and the graph navigates;
            // there is nothing left for this screen to do.
        }
    }

    /** Returns the URL for the Custom Tab, or null if the provider is off. */
    suspend fun mindlogIdUrl(create: Boolean): String? =
        if (_state.value.providers.mindlogId) authRepository.mindlogIdAuthUrl(create) else null

    fun submitPendingEmail() {
        val current = _state.value
        val token = current.pendingToken ?: return
        if (current.pendingEmail.isBlank()) return
        _state.update { it.copy(busy = true, error = null) }
        viewModelScope.launch {
            authRepository.completeMindlogId(token, current.pendingEmail.trim())
                .onFailure { e -> _state.update { it.copy(busy = false, error = e.readable()) } }
        }
    }

    private fun onOAuthCallback(callback: OAuthCallback) {
        when (callback) {
            is OAuthCallback.Tokens -> viewModelScope.launch {
                authRepository.adoptTokens(callback.accessToken, callback.refreshToken)
            }
            is OAuthCallback.NeedsEmail ->
                _state.update { it.copy(pendingToken = callback.pendingToken, busy = false) }
            is OAuthCallback.Failed ->
                _state.update { it.copy(busy = false, error = callback.reason) }
        }
    }
}

private fun Throwable.readable(): String = when (this) {
    is retrofit2.HttpException -> when (code()) {
        401 -> "Incorrect email or password."
        429 -> "Too many attempts. Try again shortly."
        else -> "The server refused the request (${code()})."
    }
    is java.io.IOException -> "Cannot reach the server."
    else -> message ?: "Something went wrong."
}

File: ios-native/Sources/FeatureAuth/LoginViewModel.swift
import CoreData
import CoreNetwork
import Foundation
import Observation

@MainActor
@Observable
public final class LoginViewModel {

    private let authRepository: AuthRepository
    private let callbackBus: OAuthCallbackBus

    public var email = ""
    public var password = ""
    public var pendingEmail = ""

    public private(set) var busy = false
    public private(set) var error: String?
    public private(set) var providers = AuthProviders()

    /// Non-nil once mindlog id returned an account without an email.
    public private(set) var pendingToken: String?

    @ObservationIgnored private var listening = false

    public init(authRepository: AuthRepository, callbackBus: OAuthCallbackBus) {
        self.authRepository = authRepository
        self.callbackBus = callbackBus
    }

    public var canSubmit: Bool {
        !busy && !email.isEmpty && !password.isEmpty
    }

    /// Called by the screen when it appears — the counterpart of the Android
    /// view model's `init`, which cannot be used here because a SwiftUI view may
    /// build its model more than once.
    public func start() {
        guard !listening else { return }
        listening = true

        // Ask the server which sign-in paths it has configured before drawing
        // any of their buttons.
        Task { providers = await authRepository.authProviders() }
        Task {
            for await callback in callbackBus.callbacks { handle(callback) }
        }
    }

    public func onFieldEdited() {
        error = nil
    }

    public func signIn() {
        guard canSubmit else { return }
        busy = true
        error = nil
        Task {
            let result = await authRepository.login(
                email: email.trimmingCharacters(in: .whitespacesAndNewlines),
                password: password
            )
            // On success the session store flips `signedIn` and the root view
            // swaps the screen; there is nothing left for this one to do.
            if case let .failure(cause) = result {
                busy = false
                error = Self.readable(cause)
            }
        }
    }

    public func submitPendingEmail() {
        guard let token = pendingToken, !pendingEmail.isEmpty else { return }
        busy = true
        error = nil
        Task {
            let result = await authRepository.completeMindlogId(
                pendingToken: token,
                email: pendingEmail.trimmingCharacters(in: .whitespacesAndNewlines)
            )
            if case let .failure(cause) = result {
                busy = false
                error = Self.readable(cause)
            }
        }
    }

    // MARK: - mindlog id

    /// Returns the URL to open, or nil if the provider is off.
    public func mindlogIdURL(create: Bool) -> URL? {
        providers.mindlogId ? authRepository.mindlogIdAuthURL(create: create) : nil
    }

    public var callbackScheme: String { authRepository.callbackScheme }

    /// The URL the authentication session came back with.
    public func handleCallback(_ url: URL) {
        callbackBus.handle(url)
    }

    /// The user dismissed the sheet, or the session failed to start.
    public func mindlogIdCancelled() {
        busy = false
    }

    private func handle(_ callback: OAuthCallback) {
        switch callback {
        case let .tokens(accessToken, refreshToken):
            authRepository.adoptTokens(accessToken: accessToken, refreshToken: refreshToken)
        case let .needsEmail(pendingToken):
            self.pendingToken = pendingToken
            busy = false
        case let .failed(reason):
            busy = false
            error = reason
        }
    }

    /// The wording the Android view model's `Throwable.readable()` produces,
    /// case for case.
    private static func readable(_ error: APIError) -> String {
        switch error {
        case let .http(status, _):
            switch status {
            case 401: "Incorrect email or password."
            case 429: "Too many attempts. Try again shortly."
            default: "The server refused the request (\(status))."
            }
        case .transport:
            "Cannot reach the server."
        case .invalidURL:
            "That server address is not usable."
        case .decoding:
            "The server sent something this version cannot read."
        }
    }
}
